languages/tolk/types/cells.mdx

```diff
@@ -0,0 +1,292 @@
+
+```tolk
+struct Demo {
+    ref1: cell          // untyped ref
+    ref2: Cell<Inner>   // typed ref
+    ref3: Cell<int256>? // maybe ref
+}
+```
+
+<Aside type="tip">
+  Yes, `point.toCell()` really gives <code>{'Cell<' + 'Point' + '>'}</code>
+</Aside>
+
+A typed cell can be assigned to `cell` implicitly.
+
+## Slices: cells opened for reading
+
+To manually read data from a cell, use `beginParse()` to get a slice:
+
+```tolk
+var s = cell.beginParse();
+```
+
+Then load data incrementally: integers, coins, sub-slices, references, etc.
+
+```tolk
+val mode = s.loadUint(8);
+val dest = s.loadAddress();
+val firstRef = s.loadRef();   // cell
+
+if (s.remainingBitsCount()) {
+   // ...
+}
+```
+
+An IDE will suggest applicable methods after a dot.
+
+## Builders: cells at the moment of writing
+
+To manually construct a cell, create a builder, write some data, and finalize this builder:
+
+```tolk
+var b = beginCell();
+b.storeUint(123, 8);
+b.storeAddress(dest);
+val c = b.endCell();
+```
+
+Since methods `storeXXX` return `self`, these calls can be chained:
+
+```tolk
+val c = beginCell()
+    .storeUint(123, 8)
+    .storeAddress(dest)
+    .endCell();
+```
+
+## How to read from a builder
+
+The only way to access already written bits is to convert a builder into a slice:
+
+```tolk
+var s = b.asSlice();
+// or (absolutely the same)
+var s = b.endCell().beginParse();
+```
+
+Constructing a cell is generally expensive in terms of gas, but `b.endCell().beginParse()` is optimized to a cheap asm instruction `BTOS` without intermediate cell creation.
+
+## Auto packing to/from cells
+
+Tolk type system is designed to avoid cumbersome manual work with slices and builders.
+Almost every practical use case can be represented with an auto-serializable structure.
+
+```tolk
+struct Something {
+    // ...
+}
+
+fun parseAndModify(c: cell): cell {
+    var smth = Something.fromCell(c);
+    // ...
+    return smth.toCell();
+}
+```
+
+Having `Cell<T>`, just call `load()` to get `T`:
+
+```tolk
+fun parsePoint(c: Cell<Point>) {
+    // same as `Point.fromCell(c)`
+    var p = c.load();
+}
+```
+
+Read a detailed article [automatic serialization](/languages/tolk/features/auto-serialization).
+
+Internally, `fromCell()` does `beginParse()` and reads data from a slice.
+
+## Auto packing to/from builders/slices
+
+A struct can be parsed not only from a cell but also from a slice:
+
+```tolk
+val smth = Something.fromSlice(s);
+```
+
+Auto-serialization works at low-level also: by analogy with `loadUint()` and others, there is a `loadAny<T>()` method:
+
+```tolk
+val smth = s.loadAny<Something>();
+// or even
+val smth: Something = s.loadAny();  // T=Something deduced
+```
+
+<Aside
+  type="caution"
+  title={"T.fromSlice(s) does not mutate the slice, but s.loadAny<T>() does"}
+>
+  <p>By analogy, a call `doSmth(s)`, does not change `s`, whereas `s.loadAddress()` shifts its internal pointer.</p>
```

Comment:
This description is not clear to me, maybe we can expand it to a separate paragraph or so